components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java

```diff
@@ -18,95 +18,226 @@
-                    " was not found ");
+        if (identityStoreConnector == null) {
+            throw new DomainException(String
+                    .format("IdentityStoreConnector %s was not found", identityStoreConnectorId));
         }
 
+        return identityStoreConnector;
     }
 
     @Override
     public Map<String, IdentityStoreConnector> getIdentityStoreConnectorMapForDomain(
-            String domainName) throws DomainManagerException {
+            String domainName) throws DomainException {
 
-        return getDomainFromName(domainName)
+        return this.getDomainFromDomainName(domainName)
                 .getIdentityStoreConnectorMap();
 
     }
 
+    @Override
+    public void addCredentialStoreConnectorToDomain(
+            CredentialStoreConnector credentialStoreConnector,
+            String domainName) throws DomainException {
+
+        Domain domain = this.getDomainFromDomainName(domainName);
+
+        domain.addCredentialStoreConnector(credentialStoreConnector);
+    }
+
+    @Override
+    public CredentialStoreConnector getCredentialStoreConnector(
+            String credentialStoreConnectorId, String domainName) throws DomainException {
+
+        Domain domain = this.getDomainFromDomainName(domainName);
+
+        CredentialStoreConnector credentialStoreConnector = domain
+                .getCredentialStoreConnectorFromId(credentialStoreConnectorId);
+
+        if (credentialStoreConnector == null) {
+            throw new DomainException(String
+                    .format("credentialStoreConnector %s was not found", credentialStoreConnectorId));
+        }
+
+        return credentialStoreConnector;
+    }
+
+    @Override
+    public Map<String, CredentialStoreConnector> getCredentialStoreConnectorMapForDomain(
+            String domainName) throws DomainException {
+
+        return this.getDomainFromDomainName(domainName)
+                .getCredentialStoreConnectorMap();
+    }
+
+    @Override
+    public Domain getDomainFromClaim(Claim claim) throws DomainException {
+
+        Set<Map.Entry<Integer, Map<String, Domain>>> set = domainPriorityToDomainMap.entrySet();
+
+        for (Map.Entry<Integer, Map<String, Domain>> entry : set) {
+
+            for (Map.Entry<String, Domain> domainMapEntry : entry.getValue().entrySet()) {
+
+                Domain domain = domainMapEntry.getValue();
+
+                if (domain.isClaimBelongToDomain(claim)) {
+
+                    return domain;
+                }
+            }
+        }
+
+        throw new DomainException(String.format("Claim URI %s do not belong to any domain", claim.getClaimURI()));
+    }
+
+    /**
+     * Get all the domains to a domain name -> domain instance map.
+     *
+     * @return Map<String, Domain> map between domain name to Domain
+     */
+    private Map<String, Domain> getAllDomainNameToDomainMappingMap() {
+
+        Map<String, Domain> domainMap = new HashMap<>();
```

Comment:
Instead of creating a new Map for each call this map should be pre-populated and saved as a class attribute.